### PR TITLE
Stuff

### PR DESCRIPTION
**Pull request by kd-11**
- Disable thread scheduler for intel CPUs
- Fix a race condition causing crashing in overlays
- Promote FIFO optimizations to stable as it seems to work perfectly. This prevents the feature from getting disabled by strict mode and allows strict mode to work with usable performance. A debug option is provided to manage this for testers/devs.
- Finish reimplementation of sys_timer_usleep. A yield implementation already exists which kinda made my previous overengineered solutions useless :(

### CODE REVIEW

**Comment by jon-reid** [comment]
2f40000

RDR - 6600k - 1070 VULKAN

Getting error on main menu
```
F {rsx::thread} class std::runtime_error thrown: Invalid location (offset=0x1859100, location=0x8)
(in file Emu\RSX\RSXThread.cpp:90)
```

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2041629/RPCS3.log.gz)

File: rpcs3/Emu/Cell/lv2/sys_timer.cpp [hunk 1]
```diff
@@ -294,14 +294,28 @@ error_code sys_timer_usleep(ppu_thread& ppu, u64 sleep_time)
...
+		u64 remaining;
+
+		lv2_obj::sleep(ppu, sleep_time);
+
+		while (sleep_time > passed)
+		{
+			remaining = sleep_time - passed;
 
-	lv2_obj::sleep(ppu, std::max<u64>(1, sleep_time));
+			if (remaining > 500)
```
Comment (pauls-gh) [inline rpcs3/Emu/Cell/lv2/sys_timer.cpp]: Add a comment about why we use 500 here.  Or use a const with a meaningful variable name.

**Comment by Ristelle** [comment]
Can Confirm that this fixes Vesperia speed issues.
However it brought in stuttering audio issues on weaker(?) PC's
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2041728/RPCS3.log.gz)

**Comment by Illynir** [comment]
Yes, on Quadcore and less the pressure is huge. Regression in performance almost everywhere and SPU options (Spu loop detection, Spu thread, Lower spu thread) behave strangely and often lead to an even stronger regression instead of improving performance.

You should consider either an automatic detection of the number of threads or an option to disable it so that it is viable on weaker PCs like i5.

**Comment by Zangetsu38** [comment]
So it is really this PR, have fixed also Motorstorm Apocalypse, libvdec needed for fix video render, before freeze on first intro, now go to ingame, with render on rsx very broken, artefact glitch and freeze (spu maybe for it) etc 
img
img

**Comment by Luffykun007** [comment]
performance is much bottom than the master.

Master Build:
capturar3
----------------------------------------------------------------------------------------------------------------
This PR:
capturar4

**Comment by Luffykun007** [comment]
performance is much bottom than the master now.

Master Build:
capturar4
-----------------------------------------------------------------------------------------------------------------
This PR:
capturar3
CPU: I5 6400
GPU: GTX 960
Ram: 8 GB
Windows 10 version 1709

**Comment by Asinin3** [comment]
For me, performance is exactly the same in Persona 5 and Raging Blast 1. I also compared other games like Uncharted 1 and Yakuza 4.

**Comment by Zangetsu38** [comment]
Normal you have ryzen, user have i5 with 4 thread and so is same like my  friend have lost big speed on lot game

**Comment by Illynir** [comment]
@Asinin3 

Performance regression will only manifest itself on those who don't have enough thread, so i3/i5, Ryzen 3/Maybe Ryzen 5, not on Ryzen 7, i7.

**Comment by Parthorisian** [comment]
**Demon souls [BLUS30443]** - have infinite loading loop before going to menu https://i.imgur.com/0kmylNv.png Log [DemonSoulz_stuckRPCS3.zip](https://github.com/RPCS3/rpcs3/files/2041931/DemonSoulz_stuckRPCS3.zip)

**Comment by kd-11** [comment]
@Parthorisian Do not mention unrelated stuff. It boots fine I checked, it hangs same way as master. Nothing has changed for it.

**Comment by stevphie** [comment]
Linux build not affected for me at least... if this help = 4 core amd 9810P...

**Comment by kd-11** [comment]
I have reverted the SPU option timing changes as lower end CPUs could not handle them. Also added a workaround for sleep management on low end CPUs. Please retest if affected. Note that this is a core change, so do not expect your original options to still be optimal, retry different options for CPU performance.

**Comment by Ristelle** [comment]
Vesperia still has audio issues. With the same settings, Audio stutters even worse. Using LLVM SPU Recompiler alleviates it by a bit.

**Comment by ruipgpinheiro** [comment]
This PR seems to have fixed (or made extremely rare) the P5 deadlocks. Haven't seen any yet.

**Comment by Luffykun007** [comment]
@ KD-11 still the same as before and with worse audio now tested with commit 3c3f943f

capturar2
-------------------------------------------------------------------------------------------------------------------
capturar3

**Comment by Illynir** [comment]
So, in terms of pure FPS it's a little better than before even if it's not as much as on the master. But for example on Persona 5 the FPS fluctuates enormously now with big falls, the audio stutter enormously, the video intro stutter also whereas before it was fluid and overall the perfs are very bad in terms of feeling, much worse than the master.

This is also the case on other games, but it is a bit on a case-by-case basis, I imagine that games using SPU on a massive scale are more affected.

Ironically with the SPU LLVM and these new timings we get better performances than on master, unfortunately very little game works with the SPU LLVM.

Clearly it will be very complicated to adapt this code as much for quadcore and less than for Ryzen 7/i7...

**Comment by kd-11** [comment]
The problem here is fairly simple, lowering the yield threshold will allow usleep to jump to 501us for every sleep duration above it on windows, so we can lower it to say, 10 and suddenly some CPUs are happy again. This is however tricky and will hide timing problems, bringing issues like emulator freezing or unpredictable performance in some situations. I'm also not comfortable with yet another config option, we already have a mess of options right now with many reports of bad performance being the result of poor understanding of the configuration options. I'll ask for other opinions on this before deciding what to do with it.

**Comment by Illynir** [comment]
I don't think I misunderstand the options and general configuration of RPCS3 given the thousands of hours of testing I've already done on this emulator.

I always test all possibilities with the "Spu thread" (Auto/1/2), with or without "Lower Spu thread", with or without "Spu loop detection", ASMJIT or SPU LLVM, etc.. I spend a lot of time on this every time to get an objective opinion.

Even if some options can help a little (notably disable everything related to SPU in fact) it is clear that overall performance is lower than on Master and that audio stutter much more.

The only combination that works better for me compared to the Master are the games that work with the SPU LLVM, which is very rare as you know.

Then I don't know what else to say and I have no interest in inventing problems that don't exist.

**Comment by kd-11** [comment]
@Illynir You misunderstand, I don't mean issues listed here, I mean on the issue tracker. Obviously you know how to use most options as do most people involved in development and testing of rpcs3. But if you see all the bogus issues we always have to close all the time on github you'll understand what I mean. (see https://github.com/RPCS3/rpcs3/issues?utf8=%E2%9C%93&q=label%3A%22Meta%3A+Invalid%22+)

As a workaround, I'll lower allow_yield to 10 usecs for all quads. Looks like 6 cores are actually not affected.

**Comment by spyropt** [comment]
i have been getting lower fps mostly on p5 since a few builds ago, since the first commit for solving tales o vesperia low fps in windows
my cpu is xeon x5675 with 6/12 threads
One thing that i noticed is that i get alot more semaphore timeouts with this pr than master

**Comment by Parthorisian** [comment]
Even if this "unrelated stuff" my own opinion [BLUS30405] is too quick+have awful soun [master video](https://streamable.com/3qbqz)
[current pr vid](https://streamable.com/u54lp)

**Comment by Illynir** [comment]
Jimits your FPS already to start and indeed it has nothing to do with this PR.

**Comment by linkhimura** [comment]
kd explains that a lot of people file incorrect issues, and not 6 hours after that, a person does that exact same thing again in this PR, lol. (sorry for the spam, but I couldn't suppress my urge to point out the irony, sorry again)

**Comment by theoldsport** [comment]
Hey, this fixes Nier performance difference between Windows/Linux!

**Comment by kd-11** [comment]
Someone with an i5 please retest this. Is performance back to normal now?

**Comment by apathyrecharge** [comment]
i still seem to be getting worse performance in P5 as of the latest commit.
Start Menu: 19-26FPS (was 30)
Schoolhallway: 11-14FPS (was ~20)
Shibuya Central Street: 11-13FPS (was ~17)

**OS:** Arch Linux 4.16.11-1
**CPU:** i5-6600k @ 4.5GHz
**GPU:** GTX 970
**Driver:** 396.24

**Comment by kd-11** [comment]
I thought this changes have no impact on linux?

**Comment by Illynir** [comment]
Full test sent privately KD, too long to be here.

Be careful people, SPU options have a strange behavior. Checking your performance on master with all by default and the same on build to KD, it's the best way to test.

**Comment by apathyrecharge** [comment]
@kd-11 i can't comment on the code-end of things since i don't program, but i've observed my performance is as expected in #4646 #4649 and the master, yet lower with this PR.

@Illynir suggestion of default SPU options actually gives me even worse performance, ~15fps in the main menu, 10-12fps in school hallway and 9-11fps in shibuya

**Comment by Luffykun007** [comment]
@kd-11 sorry but it's still the same performance reported from before,is much bottom
 than the master at the moment.

**Comment by 13xforever** [comment]
So, I've been playing P5 _a lot_ with these builds and I can say for certain that build (`173a0ad7`) has almost no semaphore timeouts/swapped textures, while `3c3f943f` with reverted timing changes is plagued by them, to the point where it's unplayable after 3-4 hours (HP/SP bars get swapped, dialogue, map, _everything_).

Just mentioning it here as was asked by Asinine.

**Comment by Illynir** [comment]
@Luffykun007 I have no problem with Dragon Ball: Raging blast with this PR on my i5 4670K.

**Comment by kd-11** [comment]
p5 corruption investigation is ongoing separately, we already know how it happens ( same timing problem as GOW3 ). The timing fixes/changes brutalize lower end CPUs to an unacceptable degree which is why I ended up partially scaling back.
As for DBZ, I have consulted other testers and they seem to not have a problem @Luffykun007. What options are you using? Provide a log file for analysis.

**Comment by kd-11** [comment]
@13xforever How about https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.5-267de418? Its between those two builds and reverts some important changes that were removed from the build that helped p5.

**Comment by Zangetsu38** [comment]
good for me i have bump 50 to 60 fps in menu P5 in asmjit and in school 20 to  26

**Comment by Luffykun007** [comment]
@kd-11  these are the settings that I'm using both in the master and in your PR.

capturar

LOG: [Log RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2046742/Log.RPCS3.zip)

**Comment by kd-11** [comment]
That is the problem. You cant use the same settings here and master since timing has changed. Use threads auto then try loop detection on and off. This change intereferes with optimal settings.

**Comment by maximilian578** [comment]
as of bb8a098  for P5 changing SPU threads number to Auto brought no observable FPS increase (still very low, somewhere between 5 and 15), however changing it to 1 brought FPS up to 24 in menu and lessened audio stutter (on main for me FPS on 1 is lower than on 2). On other spu thread counts FPS remains roughly the same as on Auto. Actually, it would be great to have a debug option to log FPS and then get some overall statistics like min/max/average. It will allow measuring performance gain/loss on different settings/prs/commits/system configurations. It would be even better to bind FPS capture to a button.

**Comment by kd-11** [comment]
You also need to test if spu priority and loop detection change anything. According to testers some situations are now very different with those options on vs off. For every regression case on i5s, it has been proven to work same master or slightly better with config changes. I'd suggest staring with everything off in cpu config and then testing each option one by one.

**Comment by kd-11** [comment]
Looks like detecting hyperthreading is harder than expected. The performance problems are likely caused by using thread scheduler on an i5

**Comment by srdjan1995** [comment]
Seems ok for me now on my Ryzen 3 1200 @ 3.9 GHz. Seems to have gone to like it was before (haven't tested all games I have), and some games might have even got a bit of a speed boost (in some parts).

**Comment by kd-11** [comment]
Removed the broken hyperthreading detection. Its a complete shitshow (thanks intel). Names also have no real meaning on intel CPUs (some i7s are dual cores, and there are xeons in there as well) and trying to detect physical CPUs is more trouble than its worth due to intel changing cpuid bits all the time. Thread scheduler is therefore disabled for intel until we find evidence that it is still useful for weaker CPUs.

**Comment by alex88510** [comment]
Just a note, some newer i5 are actually 4C/8T, like the i5-8300H that I have, which is why I didn't test it with this PR. 
Might have to keep that in mind if you decided to re-implement the hyperthreading detection for Intel.

**Comment by MSuih** [comment]
Hyperthreading detection could be implemented with GetLogicalProcessorInformation on Windows and parsing `/proc/cpuinfo` on Linux.

**Comment by jobs-git** [comment]
For high cpu load applications, hyper threading mostly does not improve performance.

**Comment by Sztefanol** [comment]
NIER got quite an improvement in speed compared to earlier builds and my i5-4440 doesn't appear to be underutilized anymore
before:

this build:

**Comment by zerkocelot** [comment]
Hi there guys, I'm very new and am not sure how to implement these changes you made to get the better performance. Can someone please help out? I don't see a way for me to implement these changes.

**Comment by zerkocelot** [comment]
@legend800 i got the build im trying to use from this site! I'd just like to know how to launch it as there is no .exe

**Comment by Zangetsu38** [comment]
@Sztefanol enable framelimit to 30 and vsync and set your framlimit in auto for all

**Comment by legend800** [comment]
@zerkocelot If there's no exe, you don't have a build. Not sure what you have, but do not spam this PR and all it's watchers for stuff like this. This is not a help forum. There's a website/FAQs, forums, discord, etc. for that kind of stuff. Do not post here again. 

Admin, can you clean this thread please?

**Comment by zerkocelot** [comment]
@legend800 sorry i got the master thing from this website https://github.com/RPCS3/rpcs3/releases , and i dont know where else i could ask since this discussion is about improvements made here. I unpacked the master release and theres no exe.

**Comment by Asinin3** [comment]
@zerkocelot That page only has the very fitst release each time we change the version number, do not download off there. https://rpcs3.net/download is what you want. For support we have our own forums and discord channel https://discord.me/RPCS3 personally I would recommend discord as it's more active. GitHub is not the place for this, any further questions belong there and I won't reply here, GitHub gets enough spam as it is.

Sidenote: it's faster to direct users to the correct place to ask for support rather than just telling them not to comment here with no alternatives provided.

**Comment by AniLeo** [comment]
Spam hidden, if necessary I'll delete

**Pull request merged**